merge fix issue #1096 that caused by incomplete deletion of image anchor.

xcosmos6 opened the pull request:
Fixed issue #1096 
It was fixed on release 0.95.4 and then merged to develop. Not sure if this would work as it seems that there are a lot more difference than the 3 lines of code that I added. Please let me know if I should have done this differently and I can re-do the request. Thank you!

mihailmacarie committed:
#1992 Load workbook without crashing even if it contains sheets without explicit relationship to worksheet part (sheet tag with empty r:id).

xcosmos6 commented:
Now I see how different this file is in you latest version. So this issue might has already been fixed in the newest release.  Please let me know if this is the case so I can close this request. Thank you!

NickNack2020 committed:
Add tests for LARGE function (#1716).

sechel committed:
Load data validations from worksheet extensions.

Data validations that reference another sheet are stored in extensions, not
directly in worksheet data validations. The loading code previous didn't load
these data validations and they were missing in the workbook.

jahav committed:
Separate properties to interna and external, so internal classes can easily acces the internal methods.

jahav committed:
Store tables in XLTables with correct type. That will allow less casting across codebase, though LINQ needs to specify type because class implements multiple IEnumerable<>.

jahav committed:
Fix a table insertion check, though it is still buggy, because it checks only origin, not all cells of future table.

jahav committed:
Convert IInsertDataReader.GetData from object to XLCellValue. That is in preparation of enumerator returning directly the data and because object must be on heap, while XLCellValue is a struct.

jahav committed:
Reader can never be null and thus check doesn't make sense. It also indicated that InsertData method can return null, which it can't.

jahav committed:
Separate a function that modifies the style of the XLCell when it has its value set. The separation will help to reuse same logic for InsertData without instatntiating cells.

Start a new paradigm for modifying styles. Originally, the only way to modify the style was to instantiate a new XLStyle (through InnerStyle) and then call setter. The new way uses a modification of a StyleValue and then sets the modified StyleValue. It only allocates something, if a new style is created, because otherwise the {Style,Alignment,Font...}Value is taken from repository.

jahav committed:
Refactor the method to insert data, so it doesn't have to materialize data into 2D array, but data can be streamed while they are being inserted.

jahav committed:
When Iinserting data, use same code for both transposed and non-transposed insert.

Transposed insert materialized data into memory, but there are two benefits of the change:
* The original code marked dirty each individual cell. New one only whole inserted area.
* Code is same for both version

jahav committed:
Remove IInsertDataReader.GetRecordsCount() method, because insertion of data is now done using streaming only.

jahav committed:
Rename IInsertDataReader.GetData to IInsertDataReader.GetRecords, which is more descriptive.

jahav committed:
Remove duplicated code for updating table fields on cell value change.

jahav committed:
In some cases, field names and cells weren't in sync. Fix and add test case.

jahav committed:
Move transposition part of a insert data to a separate function for better readability.

jahav committed:
Convert XLAddress to XLSheetPoint, because sheet point is not allocated on heap.

jahav committed:
Set formulas through slices. It is done for the future, where load of cells is done fully through slices without allocation of cells (should lead to more performance code).

jahav committed:
Separate code that sets formula of a cell to separate function for clarity.

jahav committed:
Load cells of a workbook using a streaming instead making a DOM for a each row and parsing that. In the long run, loading should be done fully through streaming. Has modest performance benefit (load decreased for 250k rows, 15 columns, from 20.5 seconds to 17.5 seconds).

jahav committed:
Add a custom equals method to the structure of a value slice, because otherwise the default equals uses a reflection to determine equality, which is slow. Records are smarter, but not available in old version of C#.

jahav committed:
Remove unused method.

jahav committed:
Add information about a performance to README

ClosedXML has significantly increased its performance and to combat no-longer accurate information, add info about performance for saving a loading large workbooks.

jahav committed:
Add documentation about how formulas are calculated.

jahav committed:
Move logic to write table part to TablePartWriter. It decreases the clutter and hopefully increases encapsulation, so one writer can optimize how to write its own part without another 4k+ LOC.

jahav committed:
Move logic to write pivot table part to PivotTablePartPartWriter. It decreases the clutter and hopefully increases encapsulation, so one writer can optimize how to write its own part without another 4k+ LOC.

jahav committed:
Move logic to write pivot table cache definition part to PivotTableCacheDefinitonPartWriter. It decreases the clutter and hopefully increases encapsulation, so one writer can optimize how to write its own part without another 4k+ LOC.

jahav committed:
Move logic to write custom file properties part to CustomFilePropertiesPartWriter. It decreases the clutter and hopefully increases encapsulation, so one writer can optimize how to write its own part without another 4k+ LOC.

jahav committed:
Convert generation of extended file properties part to static in anticipation of move to a writer.

jahav committed:
Move logic to write extended file properties part to separate writer. It decreases the clutter and hopefully increases encapsulation, so one writer can optimize how to write its own part without another 4k+ LOC.

jahav committed:
Move writing of calculation chain to a dedicated writer. Better encapsulation, more modular.

jahav committed:
Move writing of a theme part to a dedicated writer. Better encapsulation, more modular.

jahav committed:
Move writing of a comments (VML) part to a dedicated writer. Better encapsulation, more modular.

jahav committed:
Remove allocations from ColorStringParser.

jahav committed:
Remove color cache. There is no point, from performance standpoint it's not a bottleneck, no allocation few comparisons.
The key problem is that it blocks conversion of style part saving from instance method to static.

jahav committed:
Convert saving of a workbook style part from instance to static for move into a writer.

jahav committed:
Move writing of a style part to a dedicated writer. Better encapsulation, more modular.

It was necessary to provide acces to Load* methods from both loading and saving, so OpenXmlHelper for now.

jahav committed:
Convert generation of workbook part into a static method, so it can be moved to separate writer.

jahav committed:
Move saving of a workbook part from XLWorkbook_Save to a separate writer.

jahav committed:
Bump version to 0.103.0-beta

jahav committed:
Rename writer file to match the class inside.

jahav committed:
Ignore warning about file in incorrect folder according to namespace for ClosedXML.Excel.PivotTables namespace.

jahav committed:
Move reading of a pivot table definition part to a separate reader.

sbeca committed:
Save data validations to worksheet extensions

jahav committed:
Load records for pivot tables into XLPivotTableCache. Doesn't write them out yet.

The modified test fiels contained values for a shared items in a table the definition shouldn't contains (e.g. pure numeric values not found in original table definition).

jahav committed:
Simplify pivot table cache definition loading by accessing WorkbookPart directly.

jahav committed:
Simplify code by removing redundant acces to workbook part through spreadsheet document.

jahav committed:
Check that index into shared items can fit into int when value is being added, not when it is being retrieved.

jahav committed:
Add infor that only unspecified date time is permitted in pivot cache records.

jahav committed:
Remove dead code.

jahav committed:
Use pivot cache field index instead of name. The index is used to directly index other properties of a field.

Commit:
#2179 Legacy expression blank to DateTime caused NRE.

The null value later fell to the Convert.ChangeType in explicit conversion to Expression to DateTime, which caused 'Null object cannot be converted to a value type.' exception.

jahav committed:
Add statistics to pivot cache field.

Field statistics are stored in the file and are available, even if field has no records (save source data = false). Ensure the values are propertly loaded/updated with record values and saved back.

The statistics are not very useful for ClosedXML, but Excel likely uses them for optimized loadings and field structure.

jahav committed:
Make XLPivotTables enumerate XLPivotTable directly, without need for casting.

jahav committed:
Save pivot cache records to the output file.

This will ensure, that what was loaded will be saved to the output file. Originally, the cache records were lost and thus any change to the pivot table was impossible for workbooks with changed source (deleted workbook, removed column...). Thanks to round trip load/save, that is no longer the case.

The records for new pivot tables need work, they duplicate values instead of refering to the shared items in pivot cache definition, but Excel can read the records anyway, so it is a significant is progress.

jahav committed:
Pivot cache no longer depends directly on the workbook data, but uses internal fields and records instead.

TheObliterator committed:
#1742 Ensure DateTimes with a locale prefix in their format code are still correctly interpreted as DateTimes (Jonathan--V)

TheObliterator committed:
Ensure Date and Time types and formats are correctly interpreted as DateTime and TimeSpan data types

Pre-defined NumberFormatIds 18-21 and 46 should be parsed as TimeSpan values rather than DateTime values.
For other formats, Excel treats "m" immediately after "hh" or "h" or immediately before "ss" or "s" as minutes, otherwise as a month value.

Additionally this fix merges Jonathan--V's PR #1742 concerning DateTime locale prefixes.

jahav committed:
Detach XLPivotSourceReference from actual data.

The source reference might be available, it could be deleted. This allows user to load a workbook even if pivot table uses record data without available source data.

sbeca committed:
Enable nullability analysis for IXLWorksheet and IXLNamedRanges contracts + related internal classes.

jahav committed:
Update language level and use polyfill for missing functions.

The Polyfill library is a development only dependency that has *.cs sources in the .nuget directory. The *.cs files are conditionaly compiled along the project depending on the target and thus more modern version use less polyfills.

The main benefit is ability to use struct records (that require IsExternalInit attribute for init properties), though generally it just allows a modern development even for netstandard2.0.

The polyfills are internal and thus don't pollute applications consuming ClosedXML.

jahav committed:
Change the behavior of IXLWorkbook.Cell(string) to throw exception on not-found instead of null.

Generally, I think we should throw when user is asking for something and it is not found. Sending an invalid address is a definite no-no (e.g. we throw on invalid address when specified through numbers). If that is not acceptable beahvior, user should check through things like IXLNamedRanges.Contains or use IXLNamedRanges.TryGetValue.

I don't think it's sensible to have nullabel check every time at every line wehre user asks for a cell using an address (e.g. `ws.Cell("A1").Value = 4`).

jahav committed:
IXLWorkbook.Range(string) now throws exception instead of returning null.

Generally, I think we should throw when user is asking for something and it is not found. Sending an invalid address is a definite no-no (e.g. we throw on invalid address when specified through numbers). If that is not acceptable beahvior, user should check through things like IXLNamedRanges.Contains or use IXLNamedRanges.TryGetValue.

I don't think it's sensible to have nullabel check every time at every line wehre user asks for a cell using an address (e.g. `ws.Range("A1:C7").Value = 4`).

jahav committed:
IXLWorkbook.NamedRange(string) and IXLNamedRanges.NamedRange(string) now throws exception instead of returning null.

This should be more in line with the rest of API surface.

jahav committed:
#2187 Fix NRE in IXLCell.Active.

jahav committed:
Modify color parser to handle invalid colors.

Modify color parser for ST_UnsignedIntHex so it doesn't throw on invalid colors and parses colors same way as Excel.

jahav committed:
Remove duplicated logic for parsing hex digit in color parser.

jahav committed:
Fix incorrect format of predefined format 47.

The format 47 is displayed as mm:ss.0 in Excel (hours are never displayed) and the format was missing the colon.

jahav committed:
Replace magical constants with enum values.

jahav committed:
Use preferred method for checking the content of loaded files.

jahav committed:
Improve test file for detecting DateTime type cell.

The test files checked only subset of predefined formats (only 14 and 22). The predefined formats 15 (d-mmm-yy) and 16 (d-mmm) were added to the test files.

jahav committed:
Improve test file for detecting TimeSpan type cell.

The test files checked only predefined formats 21. The predefined formats 18-20 and 45-47 were added to the test file.

jahav committed:
Recognize day/months in date to number conversion even without leading zeros.

Many cultures return only formats with leading zeros from DateTimeFormat.GetAllDateTimePatterns(), thus the attempt to recognize the date is not as fuzzy as it should be, e.g. 2026/4/3 might not be recognized, because pattern si only yyyy/dd/MM. The fuzzy parser replaces MM/dd with M/d to be more inclusive.

jahav committed:
Fix ISO8601 format used during text-to-number parsing.

The DD doesn't have a meaning, only dd does. Keep two digit, because that is part of the standard pattern for ISO8601. In most cases, the correct pattern is already included thanks to GetAllDateTimePatterns, but there might be some that don't have it.

jahav committed:
Reimplement YEAR function.

The original function was legacy and didn't return correct values for various conversions (VALUE semantic). It also didn't return #NUM! error for values before 1900.

sbeca committed:
Adapt PMT function

sbeca committed:
Adapt HLOOKUP function

sbeca committed:
Implement FV and IPMT functions

jahav committed:
Use a method instead of code duplication.

jahav committed:
Use string.Join for making a string list of ranges.

jahav committed:
Split logic for saving data validations.

Data validations are stored in a "standard" place in Worksheet part, but only if they don't reference other sheets. On the other hand, if data validation references another sheet, it is placed into an extension.
Also, the data validation in standard worksheet part can't hav e children, it must have at least one.

jahav committed:
Deduplicate namespace URIs in sheet writer.

jahav committed:
Don't calculate count of data validations twice. Set the `count` attribute in data validation extensions to the correct value instead of a stale value.

jahav committed:
Bump version to 0.104.0-preview1

jahav committed:
Change source package format to snupkg.

NuGet symbol server doesn't accept .symbols.nuget, it accepts only snupkg format. Thanks to that, symbols were not published to nuget server and users didn't have access to pdb file.

Source Link metadata are embedded in pdb metadata and because there was no symbol package, no symbols were available.

jahav committed:
Remove incorrect copyright tag.

The tag contained incorrect value MIT.

jahav committed:
AppVeyor should store snupkg as an artifact.

jahav committed:
#2077 Pdb checksum in snupkg doesn't correspond to the dll in nuget and thus publishing on nuget.org fails.

To avoid, include pdb directly to nuget package.

Likely cause is Fody (https://github.com/jbevain/cecil/issues/610).

jahav committed:
Bump version to 0.104.0-preview2 so nuget will accept another test release.

jahav committed:
Set a switch for CI, so it build deterministic builds.

jahav committed:
Set a switch for CI, so it build deterministic builds.

Despite documentation, the APPVEYOR environment variable has a value APPVEYOR and not just because it is written in appveyor.yml, by default. It seems that documentation that says the value is 'True' is incorrect (or due to some obscube reason).

jahav committed:
#2202 Preserve VML part with form controls across load/save.

Form controls are rather complex structures that have source of truth in VML part connected to the sheet through LegacyDrawing tag. Was part was incorrectly removed, when sheet didn't contain any comments.

jahav committed:
#1992 Create a sheet for all sheets of a workbook, even ones without explicit relationship.

This copies Excel behavior and allows loading of workbooks from some other producers. I am not very clear what is UnsupportedSheets for (despite going through blame and codeplex), so I am keeping it for now. Likely issue Codeplex 6932

psynomorph committed:
Fixed NullReferenceException at loading workbook with empty si element

lvxiao312 committed:
Fix InvalidCastException when workbook contains a pivot table and a chart sheet.

Worksheets referenced by the workbook can be of different type and be represented by different part
types (e.g. worksheet, chartsheet or dialogsheet). Originally, the code caused InvalidCastException
when the workbook contained a chartsheet and pivot reader tried to extract info about pivot table
from it.

jahav committed:
Add a test that self-closed si tag in shared string table can be loaded without an exception and that the loaded cell contains type as in Excel (=text, not blank).

Also add a new convenience method `TestHelper.LoadAndAssert` that has an extra parameter for a sheet. It's very common to have a test file just with one sheet and without this method, we have to repeat "get single worksheet" in many places.

jahav committed:
Add a test to ensure the pivot table loader won't crash when workbook contains chartsheet.

jahav committed:
Fix an exception thrown on load of Aspose.Cells generated workbook.

The original code for reading formula values only checked null of formula.Text which could be string.Empty, which later caused index out of bounds error in GetFormula. Aspose.Cells sometimes produces such workbooks.
Move the code for reading arguments of CellIs/Expression into its respective branches and read only required formulas.

jahav committed:
Remove setters from IXLSortElements.

IXLSortElements should be immutale to ensure that the sorting specification won't change after elements have been sorted (it would lead to inconsistencies).

jahav committed:
Separate out parsing logic for getting a range sorting order.

devsikander committed:
Update README.md - link update

Fix link to standard

jahav committed:
Fix incorrect sort behavior and improve performance.

Sort didn't take into account types and the sorting was done thus incorrectly. It also used to use unstable sort. It now matches the sorting behavior of Excel (used in Dat - Sort and AutoFilter): numers, texts, logical, error and blank. Blank is always at the end, regardless of sorting order (asc/desc). Errors are not sorted.

The performance is improved by two-phase sorting: In first phase, only row indexes are sorted and we get a correct position of each row. That is done using BCL sorting algorithm O(N*log(N) * log(N) worst case because of stable sorting). After that, rows are swapped, so each row is sapped at most once O(N).

jahav committed:
Simplify code by using record readonly struct in ValueSlice.

By using the record struct feature, there is less code and it is also slightly faster (especially because of value comparison). NET likely has optimized parth for record structs.

jahav committed:
Add info about breaking change to migration doc.

jahav committed:
Remove possible multiple enumeration warning.

jahav committed:
Remove possible multiple enumeration warning.

In internal API, I will rather push materialization upward (place of usage), in case there is a loop or something like that.

jahav committed:
Autofilter - deduplicate top/bottom filter logic.

jahav committed:
Remove multiple enumeration from cell values and several calls of distinct (that sorts internally) .

jahav committed:
Remove multiple enumerations.

jahav committed:
Optimization - Get previous element without materialization of all elements.

The Distinct() asically requires a materialization, but Maxy can just store one element and check if it is maximum.

jahav committed:
Remove unused base class.

jahav committed:
Change IXLPageSetup.FirstPageNumber type to int.

The values can be negative and int is thus better.

jahav committed:
Add XML comments for IXLAutoFilter.

jahav committed:
Add XML comments for IXLFilterColumn.

jahav committed:
Move AutoFilter filter creation (many uses) to a separate function to increase encapsulation.

jahav committed:
Create a separation between regular filter and custom filter.

While they are stored in similar places, their semantic logic will be different and thus decoupling will enable easier refactoring.

jahav committed:
Change the accepted type for regular filter of AutoFilter.

The actual value of a regular filter is a wildcard, it matches the formatted text of a cell. Semantic was wrong, so it was also updated. I though about allowing just a string, but in 99% this is good enough.

jahav committed:
Fix incorrect XML documentation of IXLFilterColumn.

jahav committed:
AutoFilter rework - 2/? - fix types for custom filters (#2239)

A rework of AutoFilter feature 2/?.

* Change type of custom filters (i.e. the compare filters), because that is what autofilter should actually type-compare to. We don't want to allow any class that just implements `IComparable<T>`, because it doesn't match content of cells.
* Streamline the condition evaluation, so the code in AutoFilter is simpler. Because wildcards compare to formatted string, we pass IXLCell, through that might change in future.
* Streamline loading and saving code of autofilter filters.
* Set the culture for whole test assembly to en-US, because otherwise it is system dependent and might fail in different locales.
* The `ToString_RespectsCulture` test had incorrect attribute. The [Culture] attribute determines if test should be ran depending on current cuture, while [SetCulture] actually sets the culture.
* The `CanLoadAutoFilterWithThousandsSeparator` test was incorrect, because it assumed autofilter converts value from string in XML to a value. In reality, the cell value is formatted and compared. To fix it, the formatted string now must use current locale (which is sensible and also how Excel behaves) plus the value was a number, not currency, so the formatting library uses different properties of the cultures' NumberFormat.

Will be improved and properly covered by tests once rework is done.

jahav committed:
Allow specification of a culture for the cell value formatted string.

jahav committed:
Explicitely implement interface of IXLAutoFilter, so rest of CXML has access to internal types without casting.

jahav committed:
Refactor Top/Bottom filter, so it uses correct condition (i.e. comparison) instead of explicitely checking against a subset of values.

jahav committed:
Simplify a code for loading the top/bottom autofilter.

jahav committed:
Remove useless classes.

Fluent interface can be implemented directly in XLFilterColumn and it is also place for the filter configuration. It also deals with a problem when another filter of improper fluent usage (the second values didn't check for filter type).

jahav committed:
Use Reapply, because original code has no way of knowing whether other filter columns show/hide the row. The only way is to trigger full filter reevaluation.

jahav committed:
Move average filter to XLFilter, so all filtering logic is concentrated there.

jahav committed:
Introduce a XLFilterType.None. It's used when filter columns doesn't have any filter set.

jahav committed:
Remove compiler warnings.

jahav committed:
Enable nullability in XLFilterConnection.

jahav committed:
API review: Remove setters from filter configuration properties. The filter configuration and modification should be done only through provided methods, not through properties.

jahav committed:
Unify methods names, so adding another condition to a filter uses `Add` and changing filter type uses `Set`.

jahav committed:
Review API: Remove setters for setting configuration filter values. the only allowed way is through methods. I can't allow users to set my internal state in unpredictable manner.

jahav committed:
Add info about breaking changes.

jahav committed:
Move filters from autofilte rto individual filter columns.

The XLAutoFilter had two collections, one for columns and one for filters. They had a tenuous relationship and moving filters to XLFilterColumn, we can check that column filters satisfy requirements (e.g. one type, number of filters for same type).

jahav committed:
Specify culture for test explicitely.

This test shouldn't depedn on assembly culture (set in csproj), to be more explicit.

jahav committed:
Streamline API of AutoFilter. Add methods switch to the filter and add a new condition, other reset filter and set the condition.

jahav committed:
Simplify evaluation of one auto filter condition.

jahav committed:
Remove a method to add filter from XLAutoFilter.

From now on, filters belong purely on XLFilterColumn.

jahav committed:
Add an additional parameter `reapply` to methods

Possile perf. optimizatin, if user sets multiple filters. Before: It used to reapply filters on each addition. Now: conigure filters and then call reapply.

jahav committed:
Perf improvement, so CachedValue doesn't have to be retrieved twice from slice.

jahav committed:
Fix a bug, where dynamic filter wasn't loaded and thus didn't work.

Originally NRE, then jusk commented out. The code should also deal with changed values of cells before reapply.

jahav committed:
Fix a bug, where top10 filter wasn't loaded and thus didn't work.

Originally NRE, then jusk commented out. The code should also deal with changed values of cells before reapply.

jahav committed:
XLFIlterColumn can now refresh when a filter is added.

Deduplicates code and decrease dependency through other types.

jahav committed:
Add a test that top10 filter works after loading.

jahav committed:
Reverse condition in test for better readability.

jahav committed:
Add a check that only allowed number of filters is set.

jahav committed:
Use cached value for formated value in filters, otherwise it could cause a formula evaulation.

jahav committed:
Regular filter matches as text, not as a wildcard.

jahav committed:
Convert filter value to string only once.

jahav committed:
Merge XLFilterType.DateTimeGrouping into XLFilterType.Regular.

The file structure allows to use both at the same time (it's in the checkbox dialogue of auto filter), so this was an artificial limitation.

jahav committed:
Make sure the filter logic is same for load and API.

jahav committed:
Add tests for regular autofilter.

jahav committed:
Add tests for custom filter with number operand.

jahav committed:
Fix bug in comparison of custom filter (allowed bool and text to throw cast exception).

jahav committed:
Add tests for custom filter comparator operands.

jahav committed:
Add tests for custom filter Equal operatos and fix few bugs, to align with excel.

jahav committed:
Fix NotEqual custom filter logic to match Excel.... there is no logic. Just buggy implementation baked into most important business software ever, see xkcd 1172.

jahav committed:
Add tests for top10 filters. Add a consition, so only allowed items/percents are allowed.

jahav committed:
Add tests for dynamic filters.

jahav committed:
Fix typos.

jahav committed:
Add documentation for sorting.

jahav committed:
Add documentation for autofilter.

jahav committed:
#2220 ClosedXML is not compatible with DocumentFormat.OpenXml v3.0.0 yet. Without explicit range of accepted versions, there is no warning if user has explicitely declared dependency on both v3.0 and ClosedXML 0.102*.

jahav committed:
Update OpenXML SDK to 3.0.1.

jahav committed:
Streamline equality code for XLStyleValue.

jahav committed:
Adjust XLStyle equality to compare purely style.

Originally, it also compared container which doesn't make sense. Also, other style classes (e.g. XLFont, XLFill) also only compare style and not container.

jahav committed:
Add links for migration guides to public documentation.

jahav committed:
#2144 Fix incorrect style for cells with non-materialized style.

Originally, it used only one style from an axis with preference for rows, but combination should be done on each component individually. It di9dn't work when both column and row had a style.

jahav committed:
Update ClosedXML.Parser to v 1.0.0.

jahav committed:
Structured references are evaluated during formula evaluation, so even if table
dimenstions change, new evaluation gets up-to-date state. When str-ref is not
valid (missing table ect.), Excel returns #REF!, but for [#This Row] out of
bounds, it returns #VALUE!.

The Excel GUI allows to specify Table[[#This Row],[Col]] as a Table[@Col], but
that is presentation only. The file stores [#This Row] and parser also only
parses [#This Row], not @.

jahav committed:
Width of a table (col E) in sheet4 is now correct according to adjust to content.

The original file was incorrect, because the column 5 has been inserted to a
table, but didn't have a column set a value to a cell. Because there was no
value in cell, adjust to content didn't do its work properly. With structured
references, teh adjust to content called GetFormattedString which called Value
which called recalculatethat touched touch field names and thus are rescaned...

Whole XLTable will have to be reworked anyway, because of improper
materialization and other stuff.

jahav committed:
Table field names should be case insensitive.

OOXML in general uses case insensitive comparisons for text. Table can't have
two fields that differ in case only. This is mostly kludge so keep structured
references working, XLTable needs some serious TLC.

jahav committed:
Fix a XLSheetPoint display in debugger, it used to be `"A"12`, not it's `"A12"`.

jahav committed:
Replace regex-powered code for A1-R1C1 formula conversion with AST-based one.

The formula style conversion is really important for things like copying and other stuff, but never properly worked. This allows us to significantly simplify code and improve accuracy. Also, compiling the regex did take some tame and slowed down the startup.

jahav committed:
Rename IXLNamedRange to IXLDefinedName (#2258)

Rename NamedRange to DefinedName.

The original name didn't properly express possible content of the defined name. It can be any formula, function, lambda, constant or a range. Original properties and methods were kept with [Obsolete] attribute, though interface have been renamed.

jahav committed:
Improve defined names xml docs.

jahav committed:
Simplify the XLDefinedName ctor logic.

The end goal is to move from range based defined name to generic formula defined name.

jahav committed:
Move source of the truth for defined names to a formula.

Originally, it was in Ranges list that couldn't fully express all values of defined name.

jahav committed:
Update System.Data.SqlClient due to reported security vulnerability.

jahav committed:
Update test SDK 16.11.0 because transitive dependency Newtonsoft has reported security vulnerabilities.

Version 17 fails to build due to some assembly errors in net461..

jahav committed:
#2254 When sheet is renamed, rename it also in defined name formula .

Arjan321 committed:
Remove outdated transitive package upgrades

jahav committed:
Split read of pivot table definition into its own reader.

jahav committed:
Add nullable attributes to decrease amount of possible null value warnings.

jahav committed:
Fix ROW function, so it works properly even in array formulas.

jahav committed:
Reimplement legacy MAX function.

jahav committed:
#2221 Try to make legacy functions work with error argument.

Of course it is still buggy, but legacy functions never supported errors anyway, nor
formulas. These reasons (among others) is the reason why they had to be replaced.
The correct solution is to migrate functions to new format, but that will take years ¯\_(ツ)_/¯

jahav committed:
#836 Update sheet names in formulas and names when sheet is renamed.

jahav committed:
Update fody to 1.9.0. The older version depended on net standard 1.4 which is reported as vulnerable due to System.Net.Http 4.3.0.

jahav committed:
Implement LARGE function (#1716).

jahav committed:
Improve LARGE behavior for number-like values.

jahav committed:
Add documentation how are function evaluated with arrays and various non-scalar arguments.

jahav committed:
Add tests for HLOOKUP. The original ones were unsuitable, because HLOOKUP requires sorted values and test data didn't have that.

jahav committed:
Add an adapter for HLOOKUP to take advanatage of standard coercion.

jahav committed:
Improve signature for VLOOKUP to take advanatage of standard coercion.

jahav committed:
Modify signature of PMT, FV, IPMT to double, because it's not bool (e.g. argument "TRUE" doesn't work while "0.001" does).

jahav committed:
Fix incorrect behavior of FV for zero periods.

Neither spec, not Excel says that it returns #NUM! on zero periods.

jahav committed:
Avoid NaN as a result of PMT and IPMT.

When rate is -1, there can be NaN result due to division by zero. Excel treats anything below -1 as #NUM! error.

jahav committed:
Implement a data structure to hold pivot table close to the data structure in the Open XML.

The current approach aces complexity explosion, the amoun t of possible states in both file and representation is realy large and there is a mapping between them that shoudl do that. That's never going to work properly.

jahav committed:
Rename XLPivotAreaValues to XLPivotAreaType.

jahav committed:
Rename writer for pivot table definition to be more consistent with rest or IO classes.

jahav committed:
Load all PivotTableDefinition fields into the XLPivotTable.

XLPivotTable didn't have properties for all attributes of
PivotTableDefinition. Because of it, it wasn't able to be a "dump pipe"
and just write what it read.

XLPivotTable now contains properties for all PTD attributes and reads
them. It will facilitate full representation, so we can just write the
output without patching original XML.

jahav committed:
#2276 Reimplement COUNTA function.

COUNTA was a legacy function that didn't properly take into account
types, i.e. empty string was  interpreted as a blank.

jahav committed:
Throw exception when reference counting is broken instead of silently producing invalid file.

The SST table contains value -1 for values that have 0 references. Without explicit exception,
ClosedXML produces invalid file that can't be opened by Excel (there is a dialog asking for a repair).
It's better to throw an exception instead of failing silently. There is also a possibility of stop-gap fix
by increasing number o references.

jahav committed:
Fix a bug where cells with empty text weren't explicitely written to file.

That caused several issues, including invalid files, because SST wrote -1 as a SST index.

jahav committed:
Reimplmenet FACT functions.

jahav committed:
Reimplmenet COMBIN function.

jahav committed:
Implement BINOMDIST function.

The implementation is also aliased as BINOM.DIST function. Excel says it's more correct, but I don't see how. It's as accurate as it gets.

sbeca committed:
Optimise workbook loading by stopping unneeded invalidation

jahav committed:
Remove redundant loading of props of XLPivotTable.

The removed fields were already loaded through LoadPivotTableDefinition.

jahav committed:
Bump version to 0.104.0-preview3.

This should avoid problems with collision of custom builds and build published on nuget.org.

jahav committed:
Mark 'dataCaption' as non-nullable.

In XML, the attribute is not optional, but required and thus
must be always present. It makes no sense to have it as nullable.

jahav committed:
Add and load pivot table:location attributes.

Location attributes contain dimensions that should be loaded to behave as
"dump pipe" and thus should be loaded to fully represent the table.

Load the attributes for location into XLPivotTable. The XLPivotTable.Area
property is kind of redundant wiht Target and not kept in sync. That will
be dealt in further refactoring.

jahav committed:
Fix incorrectly loaded pivot item attribute.

DrillAcrossAttributes was used twice, the second usage was incorrect.

jahav committed:
Add a skeleton of a pivot table stream writer.

This is a part of pivot table rework. There are newly filled structures
(without any checking) and the goal is to be able to "pipe" pivot table
through, as long as it's not modified and still create valid pivot table.

This adds ability to create a empty pivot table part with empty fields, if
writer is switched in XLWorkbook_Save. Currently, switch is not done and
original writer is still used to keep code dormant while it is being worked on.

Currently, the pivot table is written with PivotTableDefinitionPartWriter
that uses OpenXML SDK structures. We want to move away from OpenXML
SDK to a more streaming solution (plus more performanct).

jahav committed:
Add missing namespace for 'item' element.

jahav committed:
Fix incorrect default 'Expanded' attribute value.

jahav committed:
Add axes elements to pivot table writer.

jahav committed:
Add filters elements to pivot table writer.

jahav committed:
Add dataFields element to pivot table writer.

jahav committed:
Add loading and writing pivot tabel styles to Writer2.

Old code has a difficulty reading style because it needs to translate
from file structures to data structures. The new structures are more
in-line with a file, leading to easier loading and saving.

This adds ability to read and write formats, so it can work with
more pivot tables in various files. The code is in Writer2 that is not
used in rest of codebase yet.

We have no special object for dxf, so use XLStyle. That makes me unhappy,
but a lot of things make me unhappy.

jahav committed:
Fix repeated attribute in pivot table field.

The original code

- didn't correctly loaded items because it didn't consider whole row/col
  items.
- writer didn't take into account that non-data item types (e.g. sum)
  always have at least one x item.

jahav committed:
Add namespace decl. at the beginning of PT part.

Originally, XmlWriter added the namespace declaration at the end of
pivotTableDefinition opening tag, add it at the beginning, so comparison
is easier.

jahav committed:
Remove unused NS declaration.

Pivot table is completely under our control and there is no use at this
time of revision or markup compatibility namespace.

jahav committed:
Add pivot table conditional formats load/save.

The pivot tables weren't able to represent conditional formatting,
this adds structures and load and save code. The saving code is in
Writer2, which is not yet enabled.

Even when Writer2 is used, it doesn't work, because it needs implicit
relationship with worksheet part conditional formatting elements.
There must be CF in the PT sheet with pivot flag and same priority.
Without it, Excel tries to repair file.

jahav committed:
Add conditional formating to pivot table.

Conditional formatting for pivot tables is separated into two places:

- Pivot table CF, that contains areas where to specify and scope.
- Sheet conditional formating that contains CF rule logic

These two places are connected through implicit relationship of sheet,
pivot flag and priority.

This commit moves CF for pivot tables from sheet directly to
XLPivotConditionalFormat and modify load/saving logic. It introduces
a LoadContext that is used to stored CF from sheet, until it is used
by pivot table loaded.

jahav committed:
Rename CF.OriginalPriority to Priority.

The value is used for sorting of CF in a sheet.

jahav committed:
Remove redundant Priority from PT cond. format.

Its value is important to establish relationship between pivot table CF
and sheet CF rule. Once it is done, the actual value is redundant,
because it is stored in format instance of pivot conditional format.

jahav committed:
Improve error message in exception.

sbeca committed:
PR feedback: change from not notifying, to loading sheets in 2 loops

sbeca committed:
Add some nullable annotations to make API slightly nicer to work with

OldBuddy committed:
Fix TimeSpan conversion

jahav committed:
Add a load test for dialog sheet

Dialog sheet is nearly obsolete feature, but there was an
InvalidCastException in 0.102.2. The problem has been fixed in 92f9fe2,
but add a test we can load worbook with dialogSheet anyway.

Pankraty committed:
Fix an edge case when ranges got unmerged when the adjacent range of the same size got deleted

henriw88 commented:
The issue still persists in version 0.97. 
We are using the .Delete property, which leads to the incomplete deletion.
This is some sample code we use:

`var pictureToDelete = workSheet.Pictures.FirstOrDefault(pic => pic.Name == "VesselGA");
            pictureToDelete?.Delete();
`
The result looks like this:
![excel picture](https://github.com/ClosedXML/ClosedXML/assets/89198178/cee2b63d-c162-4a7c-9288-1eb31be57dca)

jahav committed:
Modify pivot table data structure to be xlsx-like

Original data structure not aligned with xlsx-like and it was necessary to do
rather complicated transaltion from structure in the file to the structure in
ClosedXML during load and again from structure in ClosedXML to xlsx-like
structure during saving.

Because we don't control input files, we have to accept basically anything on
the input which means that transaltion logic would be incredible complex.

Instead, the data structures are modified to allow streamlined loading of files
and we can enable simple pass-through load-save cycle.

This change means, that basically all implemented API had to be reimplemented,
along with loader and saver. The pivot tables API is basically undocumented,
but I tried my best to keep it working (will be deleted soon).

From practical point, we control API and can reject invalid operation and we
can limit possible inputs. That is not possible with files. We have to load
any valid file.

Because of PR size, the pivot styles API wasn't reimplemented, will be done in
a sepearte PR. `XLPivotTable.AllStyleFormats` needs to be fixed and following
tests have been disabled (thus needs to be enabled):

* LoadingTests.CanLoadPivotTableWithBorder
* XLPivotTableTests.PivotTableStyleFormatsTest

jahav committed:
Improve IXLStylized documentation.

Move some XML docs from XLStylizedBase to IXLStylized and add info for the rest.

jahav committed:
Delete unused IXLStylized.Styles property.

jahav committed:
Make XLNumberFormatKey to immutable.

Mutable structs are basically always evil due to defensive copting around.

jahav committed:
Change XLAlignmentKey to readonly struct.

Mutable structs are generally evil.

jahav committed:
Remove XLAlignmentKey.TopToBottom property

The property is still used in the IXLAlignment, but it just sets text
rotation to 255 to achieve the desired result. The actual property in
the XLAlignmentKey wasn't used.

jahav committed:
Mark all properties of XLAlignment as required

This is done to make sure no property is forgotten during
initialization. Also, the XLAlignmentKey is missing ReadingOrder
property, so when we add it, all initialization places will report
an error.

jahav committed:
Change XLBorderKey to readonly struct.

jahav committed:
Change XLFillKey to readonly struct.

jahav committed:
Change XLFontKey to readonly struct.

Mutable structs are generally evil.

jahav committed:
Mark all properties of XLFontKey as required.

If we ever add another property, all instantiation places should
report an error. It's rather easy to forget one (e.g. shadow in
default key)

jahav committed:
Change XLProtectionKey to readonly struct.

jahav committed:
Change XLStyleKey to immutable struct.

xcosmos6 commented:
> The issue still persists in version 0.97. We are using the .Delete property, which leads to the incomplete deletion. This is some sample code we use:
> 
> `var pictureToDelete = workSheet.Pictures.FirstOrDefault(pic => pic.Name == "VesselGA"); pictureToDelete?.Delete(); ` The result looks like this: ![excel picture](https://private-user-images.githubusercontent.com/89198178/344199900-cee2b63d-c162-4a7c-9288-1eb31be57dca.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNjY5NTcsIm5iZiI6MTcyMTA2NjY1NywicGF0aCI6Ii84OTE5ODE3OC8zNDQxOTk5MDAtY2VlMmI2M2QtYzE2Mi00YTdjLTkyODgtMWViMzFiZTU3ZGNhLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE4MDQxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0YmIzY2I5N2E1ZmRlOGVjMzYyZDM0OWNhMjlhYzhhMDg4MjI5ZDAwMDQ2NzM3YjkyNGZkNDdmNDM0ZmNjNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lDvBKT3_waUi75Xw326hv8Ua3wN8va7YUAkST4kMhT4)

I am not sure what to on this as nobody has responded to my pull request or my question. I assume you can also add those 3 lines of code into version 0.97?

igitur commented:
Pull requests without unit tests will likely not be accepted.

xcosmos6 commented:
> Pull requests without unit tests will likely not be accepted.

I am sorry that I actually don't know how to add the test. But I only changed 3 lines of code, so, maybe you can make changes accordingly on the develop branch when you work on the code next time? 

I did pull down the latest develop branch and made the change to it too yesterday.  It's in the `ClosedXML/Excel/IO/WorksheetPartWriter.cs` file, function `GetWorksheetDom()`, in the `foreach` loop after line 1321.
`

                foreach (var removedPicture in xlPictures.Deleted)
                {
                    // Clear existing anchor of the deleted picture
                    var existingAnchor = GetAnchorFromImageId(worksheetPart.DrawingsPart, removedPicture);
                    if (existingAnchor != null)
                        worksheetPart.DrawingsPart.WorksheetDrawing.RemoveChild(existingAnchor);

                    worksheetPart.DrawingsPart.DeletePart(removedPicture);
                }
`
So, maybe you can disregard this pull request?
Thanks!

xcosmos6 committed:
fixed issue 1096/2038 on latest develop branch.

xcosmos6 committed:
resolve conflict by take local XLWorkbook_Save.cs as a whole.

xcosmos6 committed:
update system.IO.packaging.
